### ISSUE TITLE
Automatic Created and Updated columns

### ISSUE THREAD

**Issue by peachp** [supabase/supabase#379]
# Feature request

## Is your feature request related to a problem? Please describe.
No specific problem right now, but it is a very common and useful feature.

## Describe the solution you'd like
When creating a table allow to add "Created" and "Updated" columns, which will be populated by the system accordingly

## Describe alternatives you've considered
Do in client side code for each table for each INSERT / UPDATE

**Comment by awalias** [comment]
Could be a good one @joshenlim

**Comment by kiwicopple** [comment]
I think for this one we would need to add triggers - https://x-team.com/blog/automatic-timestamps-with-postgresql/

We could "auto populate" the field on the dashboard, but it would be missing values when a row is programmatically inserted (via the API). Will the trigger approach work for you @peachp ?

**Comment by peachp** [comment]
@kiwicopple I think populating it on the lowest level possible, would be great. So it won't matter whether I insert / update a row via your web GUI, API, SQL or maybe even when connected to the DB directly. Though at least via API and web GUI would be good.

**Comment by kiwicopple** [comment]
Hey @peachp , we now have the `moddatetime` extension, so you can run this

```sql

create extension if not exists moddatetime schema extensions;

-- assuming the table name is "todos"
create trigger handle_updated_at before update on todos 
  for each row execute procedure moddatetime (updated_at);
```

I'll add this to the quickstart snippets

**Comment by peachp** [comment]
@kiwicopple
I could do this for tables where I really need it, though it would be great to have a GUI where the two columns (e.g. `updated_at` and `created_at`) can be activated when creating a new table or editing it under Database.
Or maybe even allow enabling it globally for all tables!

> We could "auto populate" the field on the dashboard, but it would be missing values when a row is programmatically inserted

No no, of course it wouldn't make sense to auto-populate rows only when inserting / updating via dashboard. To clarify, when I wrote
> When creating a table allow to add "Created" and "Updated" columns

By that I meant when creating a new table under Database, like here under "Include primary key" for example:
![image](https://user-images.githubusercontent.com/47209189/103882300-808d5680-50d3-11eb-8cf3-b081c9f2da24.png)

And when you edit the table, the GUI should also show whether these two fields are being auto-populated, because it might be risky to judge only by the data itself, whether they are populated: if you run `create trigger... updated_at` then do `drop trigger... updated_at` for some reason, you would have to remember (or run some SQL to see active triggers), because you would still have `update_at` values in your data, but they will be outdated.

Anyways!
What you provided is already a good start. I guess you could make it more user-friendly at some point.

**Comment by kiwicopple** [comment]
Yeah we need much better "trigger" support in the dashboard. It's coming !

> the GUI should also show whether these two fields are being auto-populated

This one will be a bit harder if they are populated by triggers, but let us brainstorm this one. I'll convert this one to a Discussion for now since it's not a breaking issue. It will be better as the community can then search for it in the future

**Comment by peachp** [comment]
Another way would be to have a special type when you add a field. Have a look how Airtable does it.

They solved it pretty good I think: you can add or remove such fields at any time, even after you added records. Implies that they always maintain something like `created_at` and `updated_at` under the hood, so when you add fields of this type, you are just "surfacing" them, so that your pre-existing records will have it too.

For `updated_at` they go even further: you can chose to update it only when certain fields change. This allows logging / auditing, but also to build some neat business logic on top of it (which we actually do).

I imagine this could be low priority on your list because of relatively high complexity / low value (?), so no rush. But it would be great to have it at some point. This is something I've seen in a few mature database services / ORMs / frameworks.

![image](https://user-images.githubusercontent.com/47209189/104025520-acd0d200-51bc-11eb-98ef-2e8d4ab73590.png)

![image](https://user-images.githubusercontent.com/47209189/104025786-0c2ee200-51bd-11eb-8495-b6fe7682a51c.png)

**Comment by kiwicopple** [comment]
For anyone stumbling upon this, we have a small workaround. There is a script in the SQL editor called "Automatically update timestamps" which will automatically update a timestamp

<img width="470" alt="image" src="https://user-images.githubusercontent.com/10214025/114698248-78d3f280-9d51-11eb-8ee2-8a2419901931.png">

**Comment by Polyterative** [comment]
Works well for me

**Comment by surjithctly** [comment]
Just wondering how do this work for Created_at, currently it does not add column when creating new rows. 

I wanted to add a column created_at when new row is created. 

Any help?

**Comment by kiwicopple** [comment]
for that @surjithctly , you can just create a column `created_at timestamptz default now()`

![image](https://user-images.githubusercontent.com/10214025/124708705-37587b00-df2d-11eb-8dd9-5e591ba56308.png)

**Comment by kiwicopple** [comment]
We now auto-add the `created_at` column

<img width="670" alt="image" src="https://user-images.githubusercontent.com/10214025/147556464-0faade6c-232c-4ab8-a00e-6fe5b3b4065e.png">


I'll convert this one to a discussion, as i think it will be useful in the future for anyone else who wants to learn about the `updated_at` triggers

**Comment by riderx** [comment]
For everyone who need `updated_at`
Full procedure :
- go to extensions and allow `moddatetime`
- add a new column in the table you need it with, type `timestamptz`
- got to SQL editor and create a new one called `Set updated_at autoupdate` copy in it 
```sql
create trigger handle_updated_at before update on [TABLENAME] 
  for each row execute procedure moddatetime (updated_at);
 ```
 where you replace TABLENAME by the name of your table
 - run the command 
 - go to your table and do a modif after few seconds you should see the new value appear.
 
 Enjoy ❤️

**Comment by MildTomato** [comment]
Should we be looking to add this in as something that can be turned on via some simple UI ?
Rather than running SQL? 

cc @kiwicopple @saltcod

**Comment by surjithctly** [comment]
@MildTomato  Yes Please

**Comment by kiwicopple** [comment]
yup, good one
